web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts

```diff
@@ -46,74 +52,184 @@ export const useFileTransfer = (
+        name: request.filename,
+        runFileTransfer: abortController =>
+          upload(request.location, file, abortController, {
+            fileRequestId: request.requestID,
+            moderatedSessionId: request.sid,
+          }),
+      });
+    },
+    [currentUser.username, download, filesStore, upload]
+  );
+
+  // handleFileTransferUpdate is called when a FILE_TRANSFER_REQUEST event is received. This is used when
+  // we receive a new file transfer request, or when a request has been updated with an approval but its policy isn't
+  // completely approved yet. An update in this way generally means that the approver array is updated.
+  function handleFileTransferUpdate(data: FileTransferRequest) {
+    setFileTransferRequests(prevstate => {
+      // We receive the same data type when a file transfer request is created and
+      // when an update event happens. Check if we already have this request in our list. If not
+      // in our list, we add it
+      const foundRequest = prevstate.find(
+        ft => ft.requestID === data.requestID
+      );
+      if (!foundRequest) {
+        return [...prevstate, data];
+      } else {
+        return prevstate.map(ft => {
+          if (ft.requestID === data.requestID) {
+            return data;
+          }
+          return ft;
+        });
+      }
+    });
+  }
+
   useEffect(() => {
     // the tty will be init outside of this hook, so we wait until
     // it exists and then attach file transfer handlers to it
-    if (tty) {
-      tty.on(EventType.FILE_TRANSFER_REQUEST, updateFileTransferRequests);
-      tty.on(
-        EventType.FILE_TRANSFER_REQUEST_APPROVE,
-        handleFileTransferApproval
-      );
-      tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
+    if (!tty) {
+      return;
     }
-
+    tty.on(EventType.FILE_TRANSFER_REQUEST, handleFileTransferUpdate);
+    tty.on(EventType.FILE_TRANSFER_REQUEST_APPROVE, handleFileTransferApproval);
+    tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
     return () => {
-      if (tty) {
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST,
-          updateFileTransferRequests
-        );
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST_APPROVE,
-          handleFileTransferApproval
-        );
-        tty.removeListener(
-          EventType.FILE_TRANSFER_REQUEST_DENY,
-          handleFileTransferDenied
-        );
+      if (!tty) {
+        return;
       }
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST,
+        handleFileTransferUpdate
+      );
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST_APPROVE,
+        handleFileTransferApproval
+      );
+      tty.removeListener(
+        EventType.FILE_TRANSFER_REQUEST_DENY,
+        handleFileTransferDenied
+      );
     };
-  }, [tty]);
+  }, [tty, handleFileTransferDenied, handleFileTransferApproval]);
```

Comment:
After you applied that patch, I noticed that the listeners are getting re-attached after every file transfer. I managed to nail it down to the `start` function from `useFilesStore`.

This patch completely eliminates the issue:

<details>
<summary>Patch 2</summary>

```diff
diff --git a/web/packages/shared/components/FileTransfer/useFilesStore.ts b/web/packages/shared/components/FileTransfer/useFilesStore.ts
index 0f0970868a..5378c2fb2f 100644
--- a/web/packages/shared/components/FileTransfer/useFilesStore.ts
+++ b/web/packages/shared/components/FileTransfer/useFilesStore.ts
@@ -98,44 +98,6 @@ export const useFilesStore = () => {
   const [state, dispatch] = useReducer(reducer, initialState);
   const abortControllers = useRef(new Map<string, AbortController>());
 
-  const start = async (options: {
-    name: string;
-    runFileTransfer(
-      abortController: AbortController
-    ): Promise<FileTransferListeners>;
-  }) => {
-    const abortController = new AbortController();
-    const fileTransfer = await options.runFileTransfer(abortController);
-
-    if (!fileTransfer) {
-      return;
-    }
-
-    const id = new Date().getTime() + options.name;
-
-    dispatch({ type: 'add', payload: { id, name: options.name } });
-    abortControllers.current.set(id, abortController);
-
-    fileTransfer.onProgress(progress => {
-      updateTransferState(id, {
-        type: 'processing',
-        progress,
-      });
-    });
-    fileTransfer.onError(error => {
-      updateTransferState(id, {
-        type: 'error',
-        progress: undefined,
-        error,
-      });
-    });
-    fileTransfer.onComplete(() => {
-      updateTransferState(id, {
-        type: 'completed',
-      });
-    });
-  };
-
   const updateTransferState = useCallback(
     (id: string, transferState: TransferState) => {
       dispatch({ type: 'updateTransferState', payload: { id, transferState } });
@@ -143,6 +105,47 @@ export const useFilesStore = () => {
     []
   );
 
+  const start = useCallback(
+    async (options: {
+      name: string;
+      runFileTransfer(
+        abortController: AbortController
+      ): Promise<FileTransferListeners>;
+    }) => {
+      const abortController = new AbortController();
+      const fileTransfer = await options.runFileTransfer(abortController);
+
+      if (!fileTransfer) {
+        return;
+      }
+
+      const id = new Date().getTime() + options.name;
+
+      dispatch({ type: 'add', payload: { id, name: options.name } });
+      abortControllers.current.set(id, abortController);
+
+      fileTransfer.onProgress(progress => {
+        updateTransferState(id, {
+          type: 'processing',
+          progress,
+        });
+      });
+      fileTransfer.onError(error => {
+        updateTransferState(id, {
+          type: 'error',
+          progress: undefined,
+          error,
+        });
+      });
+      fileTransfer.onComplete(() => {
+        updateTransferState(id, {
+          type: 'completed',
+        });
+      });
+    },
+    [updateTransferState]
+  );
+
   const cancel = useCallback((id: string) => {
     abortControllers.current?.get(id).abort();
   }, []);
diff --git a/web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts b/web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts
index d1e4fa504b..ec5f8f844e 100644
--- a/web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts
+++ b/web/packages/teleport/src/Console/DocumentSsh/useFileTransfer.ts
@@ -51,6 +51,7 @@ export const useFileTransfer = (
   addMfaToScpUrls: boolean
 ) => {
   const { filesStore } = useFileTransferContext();
+  const { start: startTransfer } = filesStore;
   const ctx = useConsoleContext();
   const currentUser = ctx.getStoreUser();
   const [fileTransferRequests, setFileTransferRequests] = useState<
@@ -58,6 +59,7 @@ export const useFileTransfer = (
   >([]);
   const { getScpUrl, attempt: getMfaResponseAttempt } =
     useGetScpUrl(addMfaToScpUrls);
+  const { clusterId, serverId, login } = currentDoc;
 
   const download = useCallback(
     async (
@@ -65,12 +67,11 @@ export const useFileTransfer = (
       abortController: AbortController,
       moderatedSessionParams?: ModeratedSessionParams
     ) => {
-      const { clusterId, serverId, login } = currentDoc;
       const url = await getScpUrl({
         location,
-        clusterId: clusterId,
-        serverId: serverId,
-        login: login,
+        clusterId,
+        serverId,
+        login,
         filename: location,
         moderatedSessonId: moderatedSessionParams?.moderatedSessionId,
         fileTransferRequestId: moderatedSessionParams?.fileRequestId,
@@ -84,7 +85,7 @@ export const useFileTransfer = (
       }
       return getHttpFileTransferHandlers().download(url, abortController);
     },
-    [currentDoc, getScpUrl]
+    [clusterId, serverId, login, getScpUrl]
   );
 
   const upload = useCallback(
@@ -94,12 +95,11 @@ export const useFileTransfer = (
       abortController: AbortController,
       moderatedSessionParams?: ModeratedSessionParams
     ) => {
-      const { clusterId, serverId, login } = currentDoc;
       const url = await getScpUrl({
         location,
-        clusterId: clusterId,
-        serverId: serverId,
-        login: login,
+        clusterId,
+        serverId,
+        login,
         filename: file.name,
         moderatedSessonId: moderatedSessionParams?.moderatedSessionId,
         fileTransferRequestId: moderatedSessionParams?.fileRequestId,
@@ -113,7 +113,7 @@ export const useFileTransfer = (
       }
       return getHttpFileTransferHandlers().upload(url, file, abortController);
     },
-    [currentDoc, getScpUrl]
+    [clusterId, serverId, login, getScpUrl]
   );
 
   /*
@@ -141,7 +141,7 @@ export const useFileTransfer = (
       }
 
       if (request.download) {
-        return filesStore.start({
+        return startTransfer({
           name: request.location,
           runFileTransfer: abortController =>
             download(request.location, abortController, {
@@ -155,7 +155,7 @@ export const useFileTransfer = (
       if (!file) {
         throw new Error('Approved file not found for upload.');
       }
-      return filesStore.start({
+      return startTransfer({
         name: request.filename,
         runFileTransfer: abortController =>
           upload(request.location, file, abortController, {
@@ -164,7 +164,7 @@ export const useFileTransfer = (
           }),
       });
     },
-    [currentUser.username, download, filesStore, upload]
+    [currentUser.username, download, startTransfer, upload]
   );
 
   // handleFileTransferUpdate is called when a FILE_TRANSFER_REQUEST event is received. This is used when
@@ -197,9 +197,11 @@ export const useFileTransfer = (
     if (!tty) {
       return;
     }
+
     tty.on(EventType.FILE_TRANSFER_REQUEST, handleFileTransferUpdate);
     tty.on(EventType.FILE_TRANSFER_REQUEST_APPROVE, handleFileTransferApproval);
     tty.on(EventType.FILE_TRANSFER_REQUEST_DENY, handleFileTransferDenied);
+
     return () => {
       tty.removeListener(
         EventType.FILE_TRANSFER_REQUEST,
```
</details>

I moved that destructuring of currentDoc outside of the effect, so that the effect deps have only what's absolutely necessary. Though this might not be needed to fix the problem of reattaching, I don't think the doc itself gets updated that often. Still, the more precise the deps are the better.

`filesStore` should not be a dep since it's a new object on every render.